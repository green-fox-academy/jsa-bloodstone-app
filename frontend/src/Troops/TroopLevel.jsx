import React, { useState } from 'react';
import PropTypes from 'prop-types';
import {
  Animated,
  View, Image, Text,
  StyleSheet, TouchableHighlight,
} from 'react-native';
import Colors from '../common/colors';
import troopRound from '../../assets/troop/troop.png';

const styles = StyleSheet.create({
  container: {
    flexDirection: 'row',
    marginBottom: 8,
    alignItems: 'center',
  },
  roundAvatar: {
    width: 40,
    height: 40,
    marginRight: 12,
  },
  cardStyle: {
    flex: 1,
    padding: 10,
    borderRadius: 10,
    flexDirection: 'row',
    backgroundColor: Colors.tealColor,
  },
  cardText: {
    color: '#fff',
    fontSize: 16,
  },
  upgradeStyle: {
    width: 60,
    height: 40,
    justifyContent: 'center',
    alignItems: 'center',
    textAlign: 'center',
    borderTopEndRadius: 5,
    borderBottomEndRadius: 5,
    backgroundColor: Colors.redColor,
  },
  upgradeText: {
    fontSize: 8,
    color: '#fff',
  },
});

function TroopLevel({ count, level }) {
  const [showUpgradeButton, setShowUpgradeButton] = useState(false);
  const [widthOfUpgrade] = useState(new Animated.Value(0));

  function handlePress() {
    setShowUpgradeButton(!showUpgradeButton);
    Animated.timing(
      widthOfUpgrade,
      {
        toValue: showUpgradeButton ? 0 : 50,
        duration: 200,
      },
    ).start();
  }

  return (
    <View style={styles.container}>
      <TouchableHighlight
        style={styles.cardStyle}
        onPress={handlePress}
        underlayColor={Colors.tealColor}
      >
        <View style={{ flexDirection: 'row', alignItems: 'center' }}>
          <Image source={troopRound} style={styles.roundAvatar} />
          {}
          <Text style={styles.cardText}>
            {`Level ${level}: ${count} Troop${count > 1 ? 's' : ''}`}
          </Text>
        </View>
      </TouchableHighlight>
      <Animated.View style={{ ...styles.upgradeStyle, width: widthOfUpgrade }}>
        <TouchableHighlight underlayColor="transparent" onPress={() => null}>
          <Text numberOfLines={1} style={styles.upgradeText}>UPGRADE</Text>
        </TouchableHighlight>
      </Animated.View>
    </View>
  );
}

TroopLevel.propTypes = {
  level: PropTypes.number.isRequired,
  count: PropTypes.number.isRequired,
};

export default TroopLevel;
