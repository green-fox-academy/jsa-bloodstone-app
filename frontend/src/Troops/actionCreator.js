import { SERVER_URL } from 'react-native-dotenv';

export const FETCH_TROOPS_SUCCESS = 'fetchTroopsSuccess';
export const FETCH_TROOPS_REQUEST = 'fetchTroopsRequest';
export const FETCH_TROOPS_FAILURE = 'fetchTroopsFailure';

const URL = `http://${SERVER_URL}/kingdom/troops`;

export function fetchTroops() {
  return (dispatch) => {
    dispatch({ type: FETCH_TROOPS_REQUEST });
    fetch(URL)
      .then((response) => {
        if (response.status === 200) {
          return response.json();
        }
        throw new Error('An error has occurred, please try later!');
      })
      .then((response) => dispatch({ type: FETCH_TROOPS_SUCCESS, payload: response.troops }))
      .catch((error) => dispatch({ type: FETCH_TROOPS_FAILURE, payload: error }));
  };
}
