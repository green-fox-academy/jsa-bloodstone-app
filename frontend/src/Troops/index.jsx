import React, { useEffect } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import PropTypes from 'prop-types';
import {
  View, ActivityIndicator,
  StyleSheet,
} from 'react-native';
import Colors from '../common/colors';

import { fetchTroops } from './actionCreator';
import TroopInformation from './TroopInformation';
import TroopLevel from './TroopLevel';
import ErrorPopup from '../ErrorPopup';

const styles = StyleSheet.create({
  levelList: {
    paddingVertical: 10,
    flexDirection: 'column',
    backgroundColor: '#fff',
    paddingBottom: 3,
    borderRadius: 7,
  },
});

function TroopLevels({ level1TroopNum, level2TroopNum, level3TroopNum }) {
  return (
    <View style={styles.levelList}>
      {level1TroopNum !== 0 && <TroopLevel level={1} troops={level1TroopNum} />}
      {level2TroopNum !== 0 && <TroopLevel level={2} troops={level2TroopNum} />}
      {level3TroopNum !== 0 && <TroopLevel level={3} troops={level3TroopNum} />}
    </View>
  );
}

TroopLevels.propTypes = {
  level1TroopNum: PropTypes.number.isRequired,
  level2TroopNum: PropTypes.number.isRequired,
  level3TroopNum: PropTypes.number.isRequired,
};

function Troops() {
  const listOfTroops = useSelector((state) => state.troops.listOfTroops);
  const isLoading = useSelector((state) => state.troops.isLoading);
  const error = useSelector((state) => state.troops.error);
  const dispatch = useDispatch();

  useEffect(() => {
    dispatch(fetchTroops());
  }, []);

  if (error) {
    return (
      <ErrorPopup message={`Oops, ${error.message}`} />
    );
  }

  if (isLoading) {
    return (
      <ActivityIndicator size="large" color={Colors.tealColor} />
    );
  }

  const level1TroopNum = listOfTroops.filter((troop) => troop.level === 1).length;
  const level2TroopNum = listOfTroops.filter((troop) => troop.level === 2).length;
  const level3TroopNum = listOfTroops.filter((troop) => troop.level === 3).length;
  const attack = level1TroopNum + level2TroopNum * 2 + level3TroopNum * 3;
  const defence = level1TroopNum + level2TroopNum * 2 + level3TroopNum * 3;
  const sustenance = level1TroopNum + level2TroopNum + level3TroopNum;

  return (
    <View style={{ flexDirection: 'column' }}>
      <TroopInformation
        attack={attack}
        defence={defence}
        sustenance={sustenance}
      />
      <TroopLevels
        level1TroopNum={level1TroopNum}
        level2TroopNum={level2TroopNum}
        level3TroopNum={level3TroopNum}
      />
    </View>
  );
}

export default Troops;
