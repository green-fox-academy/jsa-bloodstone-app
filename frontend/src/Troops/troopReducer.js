import { FETCH_TROOPS_SUCCESS, FETCH_TROOPS_REQUEST, FETCH_TROOPS_FAILURE } from './actionCreator';

const initialState = {
  listOfTroops: [],
  isLoading: false,
  error: undefined,
};

export default function troops(state = initialState, action) {
  switch (action.type) {
    case FETCH_TROOPS_REQUEST:
      return { ...state, isLoading: true };
    case FETCH_TROOPS_SUCCESS:
      return { ...state, listOfTroops: action.payload, isLoading: false };
    case FETCH_TROOPS_FAILURE:
      return { ...state, isLoading: false, error: action.payload };
    default:
      return state;
  }
}
