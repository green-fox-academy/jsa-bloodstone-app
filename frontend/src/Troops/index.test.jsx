import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';
import { useDispatch, useSelector } from 'react-redux';

import Troops from '.';

jest.mock('react-redux');

const myTroops = [{
  countByLevel: [
    {
      level: 3,
      count: 6,
    },
    {
      level: 2,
      count: 4,
    },
  ],
  _id: '5e153c8390324235e87f2256',
  owner: '5e12cae11f7ea21cc843f669',
  attack: 26,
  defence: 26,
  hp: 10,
  battleRating: 26,
}];

describe('<Troops />', () => {
  beforeEach(() => {
    useDispatch.mockImplementation(() => () => {});
  });

  afterAll(() => {
    jest.resetModules();
  });

  it('Should render loading and match the snapshot', () => {
    const mockedState = {
      troops: {
        listOfTroops: myTroops,
        isLoading: true,
        error: undefined,
      },
      auth: {
        token: 'test token',
      },
    };
    useSelector.mockImplementation((func) => func(mockedState));
    const renderer = new ShallowRenderer();
    const tree = renderer.render(<Troops />);
    expect(tree).toMatchSnapshot();
  });

  it('Should render error and match the snapshot', () => {
    const mockedState = {
      troops: {
        listOfTroops: myTroops,
        isLoading: true,
        error: new Error('test error'),
      },
      auth: {
        token: 'test token',
      },
    };
    useSelector.mockImplementation((func) => func(mockedState));
    const renderer = new ShallowRenderer();
    const tree = renderer.render(<Troops />);
    expect(tree).toMatchSnapshot();
  });
});
