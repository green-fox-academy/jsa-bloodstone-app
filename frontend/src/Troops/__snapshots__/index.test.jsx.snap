// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Troops /> Should render error and match the snapshot 1`] = `
<ErrorPopup
  message="Oops, test error"
  title="System Error"
/>
`;

exports[`<Troops /> Should render loading and match the snapshot 1`] = `
<ActivityIndicator
  animating={true}
  color="#10978b"
  hidesWhenStopped={true}
  size="large"
/>
`;
